src/apprt/gtk/GlobalShortcuts.zig

```diff
@@ -0,0 +1,412 @@
+const GlobalShortcuts = @This();
+
+const std = @import("std");
+const Allocator = std.mem.Allocator;
+
+const gio = @import("gio");
+const glib = @import("glib");
+const gobject = @import("gobject");
+
+const App = @import("App.zig");
+const configpkg = @import("../../config.zig");
+const Binding = @import("../../input.zig").Binding;
+const key = @import("key.zig");
+
+const log = std.log.scoped(.global_shortcuts);
+const Token = [16]u8;
+
+app: *App,
+arena: std.heap.ArenaAllocator,
+dbus: *gio.DBusConnection,
+
+/// A mapping from a unique ID to an action.
+/// Currently the unique ID is simply the serialized representation of the
+/// trigger that was used for the action as triggers are unique in the keymap,
+/// but this may change in the future.
+map: std.StringHashMapUnmanaged(Binding.Action) = .{},
+
+/// The handle of the current global shortcuts portal session,
+/// as a D-Bus object path.
+handle: ?[:0]const u8 = null,
+
+/// The D-Bus signal subscription for the response signal on requests.
+/// The ID is guaranteed to be non-zero, so we can use 0 to indicate null.
+response_subscription: c_uint = 0,
+
+/// The D-Bus signal subscription for the keybind activate signal.
+/// The ID is guaranteed to be non-zero, so we can use 0 to indicate null.
+activate_subscription: c_uint = 0,
+
+pub fn init(alloc: Allocator, gio_app: *gio.Application) ?GlobalShortcuts {
+    const dbus = gio_app.getDbusConnection() orelse return null;
+
+    return .{
+        // To be initialized later
+        .app = undefined,
+        .arena = .init(alloc),
+        .dbus = dbus,
+    };
+}
+
+pub fn deinit(self: *GlobalShortcuts) void {
+    if (self.response_subscription != 0) {
+        self.dbus.signalUnsubscribe(self.response_subscription);
+        self.response_subscription = 0;
+    }
+
+    if (self.activate_subscription != 0) {
+        self.dbus.signalUnsubscribe(self.activate_subscription);
+        self.activate_subscription = 0;
+    }
+
+    if (self.handle) |handle| {
+        // Close existing session
+        self.dbus.call(
+            "org.freedesktop.portal.Desktop",
+            handle,
+            "org.freedesktop.portal.Session",
+            "Close",
+            null,
+            null,
+            .{},
+            -1,
+            null,
+            null,
+            null,
+        );
+        self.handle = null;
+    }
+    self.arena.deinit();
+}
+
+pub fn refreshSession(self: *GlobalShortcuts, app: *App) !void {
+    // Ensure we have a valid reference to the app
+    // (it was left uninitialized in `init`)
+    self.app = app;
+    self.deinit();
+
+    // Update map
+    var trigger_buf: [256]u8 = undefined;
+
+    self.map.clearRetainingCapacity();
+    var it = self.app.config.keybind.set.bindings.iterator();
```

Comment:
Note to self: create better API for this, I don't think the apprt's should be doing _this level_ of introspection. Its totally fine for this PR, but I'll get around to a better API soon, because its in the hot path of me implementing global shortcuts on macOS that don't require special permissions (that they do today).